libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2515,6 +2515,49 @@ void TypeChecker::requireLValue(Expression const& _expression)
 	_expression.annotation().lValueRequested = true;
 	_expression.accept(*this);
 
+	if (auto indexAccess = dynamic_cast<IndexAccess const*>(&_expression))
```

Comment:
This escalated to a small expression visitor now... I think I'll add a second commit which moves all these checks to the expression visitors themselves and we can decide which is better.

Reply:
Now that I looked into moving this check into the expression visitors, I'd say it's better to have the checks here after all, so I'd keep it as it is now - any objections?

Reply:
Does it make sense to turn `isLValue` into an enum?

Reply:
Not sure if that's better... we could have a ``std::function<std::string()>`` returning the reason for something not being an lvalue, but also not so nice...

If we want to go for a bit more refactoring we could merge ``lValueRequested`` and ``isLValue`` into a single ``willBeWritten`` (or something similar - we wanted to rename it at some point anyways) and have the visitors themselves emit errors, if ``willBeWritten`` is true and the expression realizes that it can't be written to... but we'd need to be careful not to miss a case (since the default for ``isLValue`` was always ``false``)... do you think that would be better?